2.0/m_rpg.cpp

```diff
@@ -0,0 +1,374 @@
+
+	CmdResult Handle(const std::vector<std::string> &parameters, User *user)
+	{
+		return NPCx::Handle(parameters, user, true);
+	}
+};
+
+/** Handle /AMBIANCE
+ */
+class CommandAmbiance : public Command
+{
+	std::string text;
+
+public:
+	CommandAmbiance(Module *parent) : Command(parent, "AMBIANCE", 2, 2)
+	{
+		this->syntax = "<channel> <text>";
+	}
+
+	CmdResult Handle(const std::vector<std::string> &parameters, User *user)
+	{
+		Channel *c = ServerInstance->FindChan(parameters[0]);
+		LocalUser *localUser = IS_LOCAL(user);
+		if (localUser)
+		{
+			if (c)
+			{
+				if (!c->HasUser(user))
+				{
+					user->WriteNumeric(ERR_NOTONCHANNEL, "%s %s :You are not on that channel!", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+				if (c->GetPrefixValue(user) < OP_VALUE)
+				{
+					user->WriteNumeric(ERR_CHANOPRIVSNEEDED, "%s %s :You're not a channel operator", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+			}
+			else
+			{
+				user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+				return CMD_FAILURE;
+			}
+		}
+		else
+			this->text = parameters[1];
+
+		/* Source is in the form of: >Ambiance<!npc@[server-name] */
+		std::string amb_source = ">Ambiance<!npc@" + ServerInstance->Config->ServerName;
+
+		c->WriteChannelWithServ(amb_source, "PRIVMSG %s :%s", c->name.c_str(), this->text.c_str());
+
+		if (localUser)
+		{
+			std::vector<std::string> params;
+			params.push_back("*");
+			params.push_back("AMBIANCE");
+			params.push_back(parameters[0]);
+			params.push_back(":" + this->text);
+			ServerInstance->PI->SendEncapsulatedData(params);
+		}
+
+		return CMD_SUCCESS;
+	}
+
+	void SetText(const std::string &newText)
+	{
+		this->text = newText;
+	}
+};
+
+/** Base class for /NARRATOR and /NARRATORA
+ */
+class Narratorx
+{
+	std::string cmdName, text;
+
+public:
+	Narratorx(const std::string &cmd) : cmdName(cmd)
+	{
+	}
+
+	CmdResult Handle(const std::vector<std::string> &parameters, User *user, bool action)
+	{
+		Channel *c = ServerInstance->FindChan(parameters[0]);
+		LocalUser *localUser = IS_LOCAL(user);
+		if (localUser)
+		{
+			if (c)
+			{
+				if (!c->HasUser(user))
+				{
+					user->WriteNumeric(ERR_NOTONCHANNEL, "%s %s :You are not on that channel!", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+				if (c->GetPrefixValue(user) < OP_VALUE)
+				{
+					user->WriteNumeric(ERR_CHANOPRIVSNEEDED, "%s %s :You're not a channel operator", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+			}
+			else
+			{
+				user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+				return CMD_FAILURE;
+			}
+		}
+		else
+			this->text = parameters[1];
+
+		/* Source is in the form of: -Narrator-!npc@[server-name] */
+		std::string narr_source = std::string("-Narrator-!npc@") + ServerInstance->Config->ServerName;
```

Comment:
How is -Narrator- a valid nickname? I cannot set that as my nick because it starts with a dash.